### PR TITLE
Remove SearchContext#current and all it's threadlocals

### PR DESCRIPTION
**Pull request by s1monw**
Today `SearchContext` expose the _current_ context as a thread local which makes any kind of sane interface design very very hard. This PR removes the thread local entirely and instead passes the relevant context anywhere needed. This simplifies state management dramatically and will allow for a much leaner `SearchContext` interface down the road

Closes https://github.com/elastic/elasticsearch/issues/19341

### CODE REVIEW

**Comment by nik9000** [comment]
This deserves more love:

```
	
        @@@@@@           @@@@@@
      @@@@@@@@@@       @@@@@@@@@@
    @@@@@@@@@@@@@@   @@@@@@@@@@@@@@
  @@@@@@@@@@@@@@@@@ @@@@@@@@@@@@@@@@@
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
  @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
   @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
    @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
      @@@@@@@@@@@@@@@@@@@@@@@@@@@
        @@@@@@@@@@@@@@@@@@@@@@@
          @@@@@@@@@@@@@@@@@@@
            @@@@@@@@@@@@@@@
              @@@@@@@@@@@
                @@@@@@@
                  @@@
                   @
```

File: core/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java [hunk 1]
```diff
@@ -406,7 +406,7 @@ public String getWriteableName() {
         } else {
             DateMathParser dateMathParser = format == null ? null : new DateMathParser(format);
             return fieldType.isFieldWithinQuery(queryRewriteContext.getIndexReader(), from, to, includeLower,
-                    includeUpper, timeZone, dateMathParser);
+                    includeUpper, timeZone, dateMathParser, (QueryShardContext) queryRewriteContext); // NOCOMMIT
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java]: can it be fixed?

**Review by jpountz — APPROVED** [review]
LGTM

**Review by nik9000 — COMMENTED** [review]
The vast majority of the changes are passing the context around and renaming a method which presumably was misnamed in the first place.

The biggest actual behavior change is in TTL which was deprecated in 2.0. Now we use the current time rather than a snapshot of the start time for displaying time remaining.

File: core/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java]: I just pushed a fix for it

**Comment by s1monw** [comment]
> The biggest actual behavior change is in TTL which was deprecated in 2.0. Now we use the current time rather than a snapshot of the start time for displaying time remaining.

it's not really important this is _only_ used for rendering purposes

**Comment by jpountz** [comment]
LGTM2

**Comment by nik9000** [comment]
> it's not really important this is only used for rendering purposes

Right.

**Review by nik9000 — APPROVED** [review]

**Pull request merged**